server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java

```diff
@@ -0,0 +1,168 @@
+            responseBuilder.notUpdated(mailboxId, SetError.builder()
+                    .type("anErrorOccurred")
+                    .description("An error occurred when updating the mailbox")
+                    .build());
+        }
+    }
+
+    private Mailbox getMailbox(String mailboxId, MailboxUpdateRequest updateRequest, MailboxSession mailboxSession) 
+            throws MailboxNotFoundException {
+
+        Optional<Mailbox> mailbox = mailboxUtils.mailboxFromMailboxId(mailboxId, mailboxSession);
+        if (!mailbox.isPresent()) {
+            throw new MailboxNotFoundException(mailboxId);
+        }
+        return mailbox.get();
+    }
+
+    private void ensureMailboxName(MailboxUpdateRequest updateRequest, MailboxSession mailboxSession) {
+        if (updateRequest.getName().isPresent()) {
+            String name = updateRequest.getName().get();
+            char pathDelimiter = mailboxSession.getPathDelimiter();
+            if (name.contains(String.valueOf(pathDelimiter))) {
+                throw new MailboxNameException(String.format("The mailbox '%s' contains an illegal character: '%c'", name, pathDelimiter));
+            }
+        }
+    }
+
+    private void ensureParent(Mailbox mailbox, MailboxUpdateRequest updateRequest, MailboxSession mailboxSession) 
+            throws MailboxException, MailboxHasChildException {
+
+        if (updateRequest.getParentId().isPresent()) {
+            String newParentId = updateRequest.getParentId().get();
+            Optional<MailboxPath> newParentMailboxPath = mailboxUtils.mailboxPathFromMailboxId(newParentId, mailboxSession);
+            if (!newParentMailboxPath.isPresent()) {
+                throw new MailboxParentNotFoundException(newParentId);
+            }
+            if (mailbox.getParentId().isPresent() && !mailbox.getParentId().get().equals(newParentId)
+                    && mailboxUtils.hasChildren(mailbox.getId(), mailboxSession)) {
+                throw new MailboxHasChildException();
+            }
+        }
+    }
+
+    @VisibleForTesting
+    MailboxPath computeNewMailboxPath(Mailbox mailbox, MailboxPath originMailboxPath, MailboxUpdateRequest updateRequest, MailboxSession mailboxSession) {
+        if (updateRequest.getParentId().isPresent()) {
+            Optional<MailboxPath> newParentMailboxPath = mailboxUtils.mailboxPathFromMailboxId(updateRequest.getParentId().get(), mailboxSession);
+            String lastName = updateRequest.getName()
+                    .orElse(getCurrentMailboxName(originMailboxPath, mailboxSession));
+            return new MailboxPath(originMailboxPath, newParentMailboxPath.get().getName() + mailboxSession.getPathDelimiter() + lastName);
+        }
+        return updateRequest.getName()
+            .map(newName -> new MailboxPath(originMailboxPath, newName))
+            .orElse(new MailboxPath(mailboxSession.getPersonalSpace(), mailboxSession.getUser().getUserName(), mailbox.getName()));
+    }
```

Comment:
Maybe you can break this function in 2 functions: one for parentId handling, other for name handling